File: sdk/python/drive9.py
"""
drive9 Python SDK

A simple Python client for drive9 API.
"""

import requests
from typing import List, Optional, Dict, Any


class Drive9Client:
    """drive9 HTTP API client."""
    
    def __init__(self, base_url: str, api_key: str):
        """
        Initialize drive9 client.
        
        Args:
            base_url: API endpoint (e.g., "https://api.drive9.ai")
            api_key: Your API key from 'drive9 create' or console
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_key}"
        })
    
    def _url(self, path: str) -> str:
        """Build full URL for a path."""
        if not path.startswith("/"):
            path = "/" + path
        return f"{self.base_url}/v1/fs{path}"
    
    def write(self, path: str, data: bytes) -> None:
        """
        Upload data to a path.
        
        Args:
            path: Remote file path (e.g., "/data/file.txt")
            data: File content as bytes
        """
        resp = self.session.put(
            self._url(path),
            data=data,
            headers={"Content-Type": "application/octet-stream"}
        )
        resp.raise_for_status()
    
    def write_file(self, path: str, local_path: str) -> None:
        """
        Upload a local file.
        
        Args:
            path: Remote file path
            local_path: Local file path to upload
        """
        with open(local_path, "rb") as f:
            self.write(path, f.read())
    
    def read(self, path: str) -> bytes:
        """
        Download a file's content.
        
        Args:
            path: Remote file path
            
        Returns:
            File content as bytes
        """
        resp = self.session.get(self._url(path))
        resp.raise_for_status()
        return resp.content
    
    def read_text(self, path: str, encoding: str = "utf-8") -> str:
        """
        Download a file as text.
        
        Args:
            path: Remote file path
            encoding: Text encoding (default: utf-8)
            
        Returns:
            File content as string
        """
        return self.read(path).decode(encoding)
    
    def list(self, path: str = "/") -> List[Dict[str, Any]]:
        """
        List directory entries.
        
        Args:
            path: Directory path (default: root)
            
        Returns:
            List of entry dicts with 'name', 'size', 'isDir' keys
        """
        resp = self.session.get(self._url(path), params={"list": ""})
        resp.raise_for_status()
        return resp.json().get("entries", [])
    
    def stat(self, path: str) -> Dict[str, Any]:
        """
        Get file/directory metadata.
        
        Args:
            path: File or directory path
            
        Returns:
            Dict with 'size', 'isDir', 'revision' keys
        """
        resp = self.session.head(self._url(path))
        resp.raise_for_status()
        return {
            "size": int(resp.headers.get("X-Drive9-Size", 0)),
            "isDir": resp.headers.get("X-Drive9-IsDir") == "true",
            "revision": int(resp.headers.get("X-Drive9-Revision", 0))
        }
    
    def copy(self, src: str, dst: str) -> None:
        """
        Copy a file (zero-copy, metadata only within drive9).
        
        Args:
            src: Source path
            dst: Destination path
        """
        resp = self.session.post(
            self._url(dst),
            params={"copy": ""},
            headers={"X-Drive9-Copy-Source": src}
        )
        resp.raise_for_status()
    
    def rename(self, src: str, dst: str) -> None:
        """
        Rename/move a file (metadata only).
        
        Args:
            src: Source path
            dst: Destination path
        """
        resp = self.session.post(
            self._url(dst),
            params={"rename": ""},
            headers={"X-Drive9-Rename-Source": src}
        )
        resp.raise_for_status()
    
    def delete(self, path: str) -> None:
        """
        Delete a file or directory.
        
        Args:
            path: Path to delete
        """
        resp = self.session.delete(self._url(path))
        resp.raise_for_status()
    
    def mkdir(self, path: str) -> None:
        """
        Create a directory.
        
        Args:
            path: Directory path
        """
        resp = self.session.post(self._url(path), params={"mkdir": ""})
        resp.raise_for_status()
    
    def grep(self, query: str, path: str = "/") -> List[Dict[str, Any]]:
        """
        Semantic/keyword search.
        
        Args:
            query: Search query
            path: Search root path
            
        Returns:
            List of matching entries
        """
        resp = self.session.get(
            self._url(path),
            params={"grep": query}
        )
        resp.raise_for_status()
        return resp.json()


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = Drive9Client(
        base_url="https://api.drive9.ai",
        api_key="your-api-key"  # Get from 'drive9 create' or console
    )
    
    # Write file
    client.write("/data/hello.txt", b"Hello, drive9!")
    print("✓ Written")
    
    # Read file
    data = client.read_text("/data/hello.txt")
    print(f"✓ Read: {data}")
    
    # List directory
    entries = client.list("/data/")
    print(f"✓ Found {len(entries)} entries")
    for e in entries:
        print(f"  - {e['name']} ({e['size']} bytes)")
    
    # Copy (zero-copy)
    client.copy("/data/hello.txt", "/data/hello-copy.txt")
    print("✓ Copied")
    
    # Rename
    client.rename("/data/hello-copy.txt", "/data/hello-renamed.txt")
    print("✓ Renamed")
    
    # Delete
    client.delete("/data/hello-renamed.txt")
    print("✓ Deleted")
    
    # Cleanup
    client.delete("/data/hello.txt")
    print("✓ Cleanup done")

File: sdk/javascript/drive9.js
/**
 * drive9 JavaScript SDK
 * 
 * A simple JavaScript/TypeScript client for drive9 API.
 */

class Drive9Client {
    /**
     * Initialize drive9 client.
     * 
     * @param {string} baseUrl - API endpoint (e.g., "https://api.drive9.ai")
     * @param {string} apiKey - Your API key from 'drive9 create' or console
     */
    constructor(baseUrl, apiKey) {
        this.baseUrl = baseUrl.replace(/\/$/, "");
        this.apiKey = apiKey;
    }

    /**
     * Build full URL for a path.
     * @private
     */
    _url(path) {
        if (!path.startsWith("/")) {
            path = "/" + path;
        }
        return `${this.baseUrl}/v1/fs${path}`;
    }

    /**
     * Make authenticated request.
     * @private
     */
    async _request(method, path, options = {}) {
        const url = this._url(path);
        const headers = {
            "Authorization": `Bearer ${this.apiKey}`,
            ...options.headers
        };

        const resp = await fetch(url, {
            method,
            headers,
            ...options
        });

        if (!resp.ok) {
            const error = await resp.text();
            throw new Error(`HTTP ${resp.status}: ${error}`);
        }

        return resp;
    }

    /**
     * Upload data to a path.
     * 
     * @param {string} path - Remote file path (e.g., "/data/file.txt")
     * @param {Buffer|Blob|Uint8Array} data - File content
     */
    async write(path, data) {
        await this._request("PUT", path, {
            body: data,
            headers: { "Content-Type": "application/octet-stream" }
        });
    }

    /**
     * Upload a local file (Node.js) or File object (browser).
     * 
     * @param {string} path - Remote file path
     * @param {File|Buffer} file - File to upload
     */
    async writeFile(path, file) {
        await this.write(path, file);
    }

    /**
     * Download a file's content.
     * 
     * @param {string} path - Remote file path
     * @returns {Promise<ArrayBuffer>}
     */
    async read(path) {
        const resp = await this._request("GET", path);
        return resp.arrayBuffer();
    }

    /**
     * Download a file as text.
     * 
     * @param {string} path - Remote file path
     * @returns {Promise<string>}
     */
    async readText(path) {
        const resp = await this._request("GET", path);
        return resp.text();
    }

    /**
     * List directory entries.
     * 
     * @param {string} path - Directory path (default: root)
     * @returns {Promise<Array<{name: string, size: number, isDir: boolean}>>}
     */
    async list(path = "/") {
        const resp = await this._request("GET", path, {
            headers: { "X-Drive9-List": "true" }
        });
        const data = await resp.json();
        return data.entries || [];
    }

    /**
     * Get file/directory metadata.
     * 
     * @param {string} path - File or directory path
     * @returns {Promise<{size: number, isDir: boolean, revision: number}>}
     */
    async stat(path) {
        const resp = await this._request("HEAD", path);
        return {
            size: parseInt(resp.headers.get("X-Drive9-Size") || "0"),
            isDir: resp.headers.get("X-Drive9-IsDir") === "true",
            revision: parseInt(resp.headers.get("X-Drive9-Revision") || "0")
        };
    }

    /**
     * Copy a file (zero-copy within drive9).
     * 
     * @param {string} src - Source path
     * @param {string} dst - Destination path
     */
    async copy(src, dst) {
        await this._request("POST", dst, {
            headers: { "X-Drive9-Copy-Source": src }
        });
    }

    /**
     * Rename/move a file (metadata only).
     * 
     * @param {string} src - Source path
     * @param {string} dst - Destination path
     */
    async rename(src, dst) {
        await this._request("POST", dst, {
            headers: { "X-Drive9-Rename-Source": src }
        });
    }

    /**
     * Delete a file or directory.
     * 
     * @param {string} path - Path to delete
     */
    async delete(path) {
        await this._request("DELETE", path);
    }

    /**
     * Create a directory.
     * 
     * @param {string} path - Directory path
     */
    async mkdir(path) {
        await this._request("POST", path, {
            headers: { "X-Drive9-Mkdir": "true" }
        });
    }

    /**
     * Semantic/keyword search.
     * 
     * @param {string} query - Search query
     * @param {string} path - Search root path
     * @returns {Promise<Array>}
     */
    async grep(query, path = "/") {
        const url = new URL(this._url(path));
        url.searchParams.append("grep", query);
        
        const resp = await fetch(url, {
            headers: { "Authorization": `Bearer ${this.apiKey}` }
        });
        
        if (!resp.ok) {
            throw new Error(`HTTP ${resp.status}`);
        }
        return resp.json();
    }
}

// Example usage
async function example() {
    // Initialize client
    const client = new Drive9Client(
        "https://api.drive9.ai",
        "your-api-key"  // Get from 'drive9 create' or console
    );

    try {
        // Write file
        await client.write("/data/hello.txt", new TextEncoder().encode("Hello, drive9!"));
        console.log("✓ Written");

        // Read file
        const text = await client.readText("/data/hello.txt");
        console.log("✓ Read:", text);

        // List directory
        const entries = await client.list("/data/");
        console.log("✓ Found", entries.length, "entries");
        for (const e of entries) {
            console.log(`  - ${e.name} (${e.size} bytes)`);
        }

        // Copy (zero-copy)
        await client.copy("/data/hello.txt", "/data/hello-copy.txt");
        console.log("✓ Copied");

        // Rename
        await client.rename("/data/hello-copy.txt", "/data/hello-renamed.txt");
        console.log("✓ Renamed");

        // Delete
        await client.delete("/data/hello-renamed.txt");
        console.log("✓ Deleted");

        // Cleanup
        await client.delete("/data/hello.txt");
        console.log("✓ Cleanup done");

    } catch (err) {
        console.error("Error:", err.message);
    }
}

// Run example if in Node.js
if (typeof window === "undefined") {
    example();
}

// Export for both CommonJS and ES modules
if (typeof module !== "undefined" && module.exports) {
    module.exports = { Drive9Client };
}
